src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,32 +852,56 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
                 return 0;
             }
 
-            fixed (char* pSource = source)
+            // according to ICU User Guide: 
+            // "The overall performance is poorer than if the function is called with a zero output buffer"
+            // "A recommended size for the temporary buffer is four times the length of the longest string processed"
+            int recommendedSize = source.Length * 4; 
+
+            // "Call ucol_getSortKey() to find out how big the sort key buffer should be, and fill in the temporary buffer at the same time"
+            if (TryGetSortKey(source, options, recommendedSize, out int actualSortKeyLength, out int hash))
+            {
+                return hash; // hot path, the buffer was big enough and we don't need to call the method one more time
+            }
+            
+            // "If the temporary buffer is too small, allocate or reallocate more space for in anoverflow buffer to handle the overflow situations."
+            if (TryGetSortKey(source, options, actualSortKeyLength, out _, out hash))
             {
-                int sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, null, 0, options);
+                return hash;
+            }
 
+            throw new ArgumentException(SR.Arg_ExternalException);
+        }
+        
+        private unsafe bool TryGetSortKey(ReadOnlySpan<char> source, CompareOptions options, int requestedSortKeyLength, out int actualSortKeyLength, out int hash)
+        {
+            fixed (char* pSource = source)
+            {
                 byte[]? borrowedArr = null;
-                Span<byte> span = sortKeyLength <= 512 ?
-                    stackalloc byte[512] :
-                    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
-
+                Span<byte> span = requestedSortKeyLength <= 512 ?
```

Comment:
`(uint)requestedSortKeyLength <= 512` - helps protect against negative inputs being passed to method.

Reply:
The calling method guarantees that, I added `Debug.Assert(requestedSortKeyLength > 0);` to make it explicit